Fix/issue 1512

iparask opened the pull request:
This pull request address issue #1512 based on the discussion on that ticket.

andre-merzky commented on examples/05_unit_input_data.py:
```diff
@@ -35,7 +35,7 @@
     # Create a new session. No need to try/except this: if session creation
     # fails, there is not much we can do anyways...
-    session = rp.Session()
+    session = rp.Session(uid='TarTest')
```
Please don't hardcode session names in the examples - this will fail on the second execution.

iparask replied:
Done

andre-merzky commented on examples/05_unit_input_data.py:
```diff
@@ -92,12 +94,15 @@
+                                  {'source': 'client:///input2.dat', 
+                                   'target': 'unit:///input2.dat',
+                                   'action': rp.TARBALL}]
 
```
Wouldn't it make sense to add an additional example?  I would think we should keep the original example on the canonical staging (otherwise we'll need to explain what  TARBALL means...)

iparask replied:
Done

andre-merzky commented on src/radical/pilot/agent/staging_input/default.py:
```diff
@@ -103,7 +104,7 @@ def _handle_unit(self, unit, actionables):
         #       class definition.
         sandbox = unit['unit_sandbox']
-
+        
```
why?  :-)

iparask replied:
I think I had something there and I deleted it.

iparask replied:
done

andre-merzky commented on src/radical/pilot/agent/staging_input/default.py:
```diff
@@ -199,7 +200,13 @@ def _handle_unit(self, unit, actionables):
+            elif action == rpc.TARBALL:
+                # If somethig was staged via the tarball method, the tarball is
+                # extracted and then removed from the unit folder.
+                tar = tarfile.open(tgtdir+'/'+uid+'.tar')
```
please use spaces around `+` (PEP.8).

iparask replied:
done

andre-merzky commented on src/radical/pilot/agent/staging_input/default.py:
```diff
@@ -199,7 +200,13 @@ def _handle_unit(self, unit, actionables):
+            elif action == rpc.TARBALL:
+                # If somethig was staged via the tarball method, the tarball is
+                # extracted and then removed from the unit folder.
+                tar = tarfile.open(tgtdir+'/'+uid+'.tar')
```
Per the `if` above, you can get `client://` type URLs here - do those need expansion?

iparask replied:
I do not think there is any need to check about `client:///`. The `if` in line 161 makes sure that whatever passes does not have `client://` at the target. Am I right?

I have a feeling that either I am missing something major,

andre-merzky commented on src/radical/pilot/agent/staging_input/default.py:
```diff
@@ -199,7 +200,13 @@ def _handle_unit(self, unit, actionables):
+                tar = tarfile.open(tgtdir+'/'+uid+'.tar')
+                tar.extractall(path=tgtdir)
+                tar.close()
+                os.remove(tgtdir+'/'+uid+'.tar')
```
You may want to keep the tarballs if `RADICAL_DEBUG` is set?

iparask replied:
Are the files the user transferred enough? I agree it needs a debug message to show what is happening.

andre-merzky commented on src/radical/pilot/umgr/staging_input/default.py:
```diff
@@ -325,26 +325,58 @@ def _handle_unit(self, unit, actionables):
+                removables.append(sd)
+                sid = sd.get('uid')
+            elif action == rpc.TRANSFER:
+                self._prof.prof('staging_in_start', uid=uid, msg=did)
```
You moved the profile entry into the `elif` - now the event is  missing for the tarballs.

iparask replied:
Yes and also I need to create a new id for this staging

andre-merzky commented on src/radical/pilot/umgr/staging_input/default.py:
```diff
@@ -325,26 +325,58 @@ def _handle_unit(self, unit, actionables):
+                # and append the directive to the removables.
 
-            self._prof.prof('staging_in_start', uid=uid, msg=did)
+                if not os.path.isfile(uid+'.tar'):
```
please use spaces around `+`

andre-merzky commented on src/radical/pilot/umgr/staging_input/default.py:
```diff
@@ -325,26 +325,58 @@ def _handle_unit(self, unit, actionables):
-            self._prof.prof('staging_in_start', uid=uid, msg=did)
+                if not os.path.isfile(uid+'.tar'):
+                    tar_file = tarfile.open(uid+'.tar','w')
+                tar_file.add(src.split('client:///')[1],arcname=tgt.split('unit:///')[1])
```
Are you *sure* that the src and tgt specs match that pattern here?  Maybe use URL parsing and then `url.path`?  This differs from how the `elif` branch handles the URLs.

andre-merzky commented on src/radical/pilot/umgr/staging_input/default.py:
```diff
@@ -325,26 +325,58 @@ def _handle_unit(self, unit, actionables):
+                    tar_file = tarfile.open(uid+'.tar','w')
+                tar_file.add(src.split('client:///')[1],arcname=tgt.split('unit:///')[1])
+                removables.append(sd)
+                sid = sd.get('uid')
```
this is the same as `did` above?

andre-merzky commented on src/radical/pilot/umgr/staging_input/default.py:
```diff
@@ -325,26 +325,58 @@ def _handle_unit(self, unit, actionables):
+                self._prof.prof('staging_in_stop', uid=uid, msg=did)
+
+
+        if os.path.isfile(uid+'.tar'):
```
Wouldn't it make sense to check if the `tar_file` object exists instead of hitting the file system?

andre-merzky commented on src/radical/pilot/umgr/staging_input/default.py:
```diff
@@ -325,26 +325,58 @@ def _handle_unit(self, unit, actionables):
+            # this type of transfer was executed.
+            tar_file.close()
+
+            self._prof.prof('staging_in_start', uid=uid, msg=did)
```
Ah, here it is.  So the transfer ID is then the did of the last staging directive.  Hmm, Lets discuss this, I am not sure what analytics will think if some SDs do not have any profile events at all.  The files get transferred after all...

andre-merzky commented on src/radical/pilot/umgr/staging_input/default.py:
```diff
@@ -325,26 +325,58 @@ def _handle_unit(self, unit, actionables):
+            tgt = complete_url(uid+'.tar', tgt_context, self._log)
 
-            saga_dir.copy(src, tgt, flags=copy_flags)
+            saga_dir.copy(src, tgt, flags=rs.filesystem.CREATE_PARENTS)
```
You hardcode that flag here - that is different from the TRANSFER case.

andre-merzky commented on src/radical/pilot/umgr/staging_input/default.py:
```diff
@@ -325,26 +325,58 @@ def _handle_unit(self, unit, actionables):
+            unit['description']['input_staging'].append({'uid': sid, 
+                                                         'priority': 0, 
+                                                         'source': 'client:///'+uid+'.tar', 
+                                                         'flags': ['CreateParents', 'SkipFailed'], 
```
please use `DEFAULT_FLAGS`

andre-merzky commented on src/radical/pilot/umgr/staging_input/default.py:
```diff
@@ -325,26 +325,58 @@ def _handle_unit(self, unit, actionables):
+                                                         'priority': 0, 
+                                                         'source': 'client:///'+uid+'.tar', 
+                                                         'flags': ['CreateParents', 'SkipFailed'], 
+                                                         'action': 'Tarball',
```
please use the constant.

andre-merzky commented on tests/issue_1512/run_test.sh:
```diff
@@ -0,0 +1,44 @@
+#!/bin/bash
+
+for jj in {1,2,4,8,16,32,64,128,256,512,1024,2048}; do
```
Why the array syntax?  This works just as well and is simpler:
```
$ for i in 1 2 3; do echo $i; done
1
2
3
```

Doesn't matter really - but in general you may want to use POSIX where bash specific features are not needed, to keep scripts portable.

andre-merzky commented on tests/issue_1512/TarCpUntar.py:
```diff
@@ -0,0 +1,27 @@
+import os
```
Please apply the coding guidelines also to tests :-)
More importantly,  how is that code related to testing the #1512?   It looks like you test the tarfile module - that is not our business...

andre-merzky commented on tests/issue_1512/randomFiles.py:
```diff
@@ -0,0 +1,28 @@
+import string
```
FYI, this is a much quicker and simpler way to create files of a certain size in a shell script:
```
$ time head -c $((1024*1024)) /dev/urandom > /tmp/t.dat
r:0m0.012s  u:0m0.001s  s:0m0.011s
```

andre-merzky commented on tests/issue_1512/run_test.sh:
```diff
@@ -0,0 +1,44 @@
+    echo '---------- Half Meg ----------'
+    echo 'Time'>run_half_$jj.csv
+    for ii in {1..10}; do
+            python transfer_files.py halfMeg wrangler.tacc.utexas.edu <path> >>run_half_$jj.csv 2>>log.log
```
I am missing something - what is `<path>` here?  `> >>` looks strange to me, too.  Does this run?

andre-merzky commented on tests/issue_1512/transfer_files.py:
```diff
@@ -0,0 +1,73 @@
+import os
```
please apply coding guidelines.

andre-merzky commented on tests/issue_1512/transfer_files.py:
```diff
@@ -0,0 +1,73 @@
+    parser.add_argument('--tar',help='Transfer without tarball or with',default=False,action="store_true")
+    args = parser.parse_args()
+    
+    TansferFiles = glob.glob(args.srcfolder+'/*')
```
typo (T*r*ansfer)

iparask commented:
Hello @andre-merzky , I am not sure how to trigger a new review request. I believe I addressed most of your comments. Also Srinivas has pushed the unit tests. Can you please take another look?

Thank you

andre-merzky commented on examples/05_unit_input_data.py:
```diff
@@ -92,12 +92,12 @@
+            cud.input_staging  = [{'source': 'client:///input.dat', 
+                                   'target': 'unit:///input.dat',
+                                   'action': rp.TRANSFER}]
 
```
Wait, why is this a list now?  I am missing something...

andre-merzky commented on src/radical/pilot/umgr/staging_input/default.py:
```diff
@@ -325,26 +324,61 @@ def _handle_unit(self, unit, actionables):
+                tgt = complete_url(tgt, tgt_context, self._log)
+
+                saga_dir.copy(src, tgt, flags=flags)
+                self._prof.prof('staging_in_stop', uid=uid, msg=did)
```
This event is duplicated for this uid/did now.

andre-merzky commented:
Please run at least one remote test on XSEDE.

andre-merzky commented:
Guys, help me out - how did `tests/test_tarball_umgr.py` actually work when the `saga.fs.Directory.copy` call was `mock`'ed away?  I don't get it...

andre-merzky commented:
Ah, I do get it now (1 min after last comment :-) : you mask the ENOTDIR exception which happens when passing a URL to the shutil copy...  Man, this took me ages...

Leaving SAGA in place to do the actual thing works.  I would argue that one should only mock those parts which interfere with running the test, really.  Otherwise we don't really test much, do we?  @itomaldonado , any opinion?

itomaldonado commented:
@andre-merzky I wanted to comment on your last point. I suggested them not testing the saga layer because this test specifically isn't looking at how saga does a transfer. From this module's perspective one the saga function returns successful, the assumption is that the file made it to its destination. We can remove the dependency of Saga (and having to set-up a transfer, ssh setups etc.) by mocking the function call. You can also mock saga to return errors (i.e. raise exceptions) and have unit tests covering the scenarios when that function call fails, but we don't really care about actually testing the transfer because that is not what this test is about.

the assumption is that the saga function was tested separately.

Hopefully that makes sense?
